### ISSUE TITLE
Trying to rename field to `@timestamp` leads to duplicated field.

### ISSUE THREAD

**Issue by adriansr** [elastic/beats#7351]
Using the rename processor to rename a field to `@timestamp`, as an attempt to override it, I ended up with an event that has 2 `@timestamp` fields and fails to be indexed into ES.

Adding a drop_fields processor to get rid of `@timestamp` first doesn't change the outcome.

For confirmed bugs, please report:
- Version: 7.0.0-alpha
- Operating System: 
- Discuss Forum URL: Trying to find an answer for https://discuss.elastic.co/t/setting-timestamp-in-filebeat/136224
- Steps to Reproduce: 

filebeat.yml:
```
filebeat.inputs:
- type: log
  enabled: true
  paths:
    - ${path.config}/test.log
output.elasticsearch:
  hosts: ["localhost:9200"]
processors:
    - drop_fields: # This is optional, doesn't make a difference
        fields: ['@timestamp']
    - rename:
        fields:
            - from: message
              to: '@timestamp'
        fail_on_error: true
```

test.log (just wanted a timestamp at a known field to experiment):
```
2018-06-17 21:39:14.779
```

logs from filebeat:
```
2018-06-18T14:59:23.142+0700	INFO	log/harvester.go:228	Harvester started for file: /Users/adrian/go/src/github.com/elastic/beats/filebeat/test.log
2018-06-18T14:59:23.142+0700	DEBUG	[registrar]	registrar/registrar.go:284	Registry file updated. 1 states written.
2018-06-18T14:59:23.143+0700	DEBUG	[filter]	pipeline/processor.go:157	fail to apply processor global{drop_fields=@timestamp, rename=[{From:message To:@timestamp}]}: key not found
2018-06-18T14:59:23.143+0700	DEBUG	[publish]	pipeline/processor.go:291	Publish event: {
  "@timestamp": "2018-06-18T07:59:23.143Z",
  "@metadata": {
    "beat": "filebeat",
    "type": "doc",
    "version": "7.0.0-alpha1"
  },
  "source": "/Users/adrian/go/src/github.com/elastic/beats/filebeat/test.log",
  "offset": 0,
  "prospector": {
    "type": "log"
  },
  "input": {
    "type": "log"
  },
  "beat": {
    "name": "Adrian-Elastic.true.true",
    "hostname": "Adrian-Elastic.true.true",
    "version": "7.0.0-alpha1"
  },
  "host": {
    "name": "Adrian-Elastic.true.true"
  },
  "@timestamp": "2018-06-17 21:39:14.779"
}
2018-06-18T14:59:23.143+0700	DEBUG	[harvester]	log/log.go:85	End of file reached: /Users/adrian/go/src/github.com/elastic/beats/filebeat/test.log; Backoff now.
2018-06-18T14:59:24.148+0700	DEBUG	[harvester]	log/log.go:85	End of file reached: /Users/adrian/go/src/github.com/elastic/beats/filebeat/test.log; Backoff now.
2018-06-18T14:59:24.148+0700	DEBUG	[elasticsearch]	elasticsearch/client.go:666	ES Ping(url=http://localhost:9200)
2018-06-18T14:59:24.163+0700	DEBUG	[elasticsearch]	elasticsearch/client.go:689	Ping status code: 200
2018-06-18T14:59:24.163+0700	INFO	elasticsearch/client.go:690	Connected to Elasticsearch version 7.0.0-alpha1
2018-06-18T14:59:24.163+0700	DEBUG	[elasticsearch]	elasticsearch/client.go:708	HEAD http://localhost:9200/_template/filebeat-7.0.0-alpha1  <nil>
2018-06-18T14:59:24.165+0700	INFO	template/load.go:112	Template already exists and will not be overwritten.
2018-06-18T14:59:24.168+0700	DEBUG	[elasticsearch]	elasticsearch/client.go:303	PublishEvents: 1 events have been published to elasticsearch in 2.567835ms.
2018-06-18T14:59:24.168+0700	WARN	elasticsearch/client.go:502	Cannot index event publisher.Event{Content:beat.Event{Timestamp:time.Time{wall:0xbec1f776c88636b4, ext:68073566, loc:(*time.Location)(0x52b21a0)}, Meta:common.MapStr(nil), Fields:common.MapStr{"@timestamp":"2018-06-17 21:39:14.779", "source":"/Users/adrian/go/src/github.com/elastic/beats/filebeat/test.log", "offset":0, "prospector":common.MapStr{"type":"log"}, "input":common.MapStr{"type":"log"}, "beat":common.MapStr{"name":"Adrian-Elastic.true.true", "hostname":"Adrian-Elastic.true.true", "version":"7.0.0-alpha1"}, "host":common.MapStr{"name":"Adrian-Elastic.true.true"}}, Private:file.State{Id:"", Finished:false, Fileinfo:(*os.fileStat)(0xc42036a1a0), Source:"/Users/adrian/go/src/github.com/elastic/beats/filebeat/test.log", Offset:24, Timestamp:time.Time{wall:0xbec1f776c87536b9, ext:66959459, loc:(*time.Location)(0x52b21a0)}, TTL:-1, Type:"log", Meta:map[string]string{}, FileStateOS:file.StateOS{Inode:0xf1f185, Device:0x1000004}}}, Flags:0x1} (status=400): {"type":"parse_exception","reason":"Failed to parse content to map","caused_by":{"type":"json_parse_exception","reason":"Duplicate field '@timestamp'\n at [Source: org.elasticsearch.transport.netty4.ByteBufStreamInput@44fd3b57; line: 1, column: 97]"}}
```

**Comment by urso** [comment]
The `@timestamp` field is internal of the `beat.Event`. Most processor operate on fields only. The JSON parsers have some logic to parse `@timestamp` into the event. This should be generalized, so we can add/remove fields more easily to `beat.Event`, not `common.MapStr`.

**Comment by andrewkroh** [comment]
I get a similar problem when trying to add my own fields to `@metadata` to pass them to Logstash. I end up with two sets of `@metadata`.

```yaml
fields_under_root: true
fields:
  '@metadata.ls_pipeline': mypipeline
```

```json
  "@metadata": {
    "beat": "filebeat",
    "type": "doc",
    "version": "7.0.0-alpha1",
    "pipeline": "filebeat-7.0.0-alpha1-system-syslog-pipeline"
  },
  "@metadata": {
    "ls_pipeline": "mypipeline"
  },
```

**Comment by urso** [comment]
the `@metadata` and `@timestamp` fields are special `beat.Event` fields. The processors operate on the `Fields` only. The `rename` processor must be updated to take the full event structure into account. See json decoding processor, which uses `event.PutValue`.
Trying to move a `@metadate` field to the top-level event might also fail.

Note: `@timestamp` is tricky, as it must be parsed.

**Comment by sovetov** [comment]
It would be very nice if it was possible to operate on these special fields.

Currently, there is no straightforward way to to capture when message were received by Logstash if it has `@timestamp` field. The only option I know is [to use `ruby` filter](https://discuss.elastic.co/t/timestamps-precision-in-beats/154646/11?u=george_sovetov).

**Comment by dennybaa** [comment]
Hi there,
so what is the way out? my timestamp is already in ISO8601 and nothing should be parsed. What I need is just to rename `timestamp` -> `@timestamp` which DOES NOT WORK.

Will you suggest to use an ingest filter? Yes, it can be a workaround. But it would be nice is to hear that this a real issue and it could be resolved in the future versions...

**Comment by dennybaa** [comment]
oops, sorry, noticed the bug label :)

**Comment by dennybaa** [comment]
PS. prior dropping of the `@timestamp` obliviously does not mitigate renaming.

**Comment by nitinpatil1992** [comment]
Hello, We are facing the same issue but not for `@timestamp`. 
<img width="1057" alt="Screen Shot 2019-06-05 at 15 01 08" src="https://user-images.githubusercontent.com/45567409/58958317-f1a3b400-87a2-11e9-8d02-0c6c0808e44c.png">

Here is our configucation:
```
- rename:
    fields:
    - from: "docker.container.labels.service.name"
      to: "@service-name"
    - from: "docker.container.labels.service.environment"
      to: "@service-environment"
    - from: "docker.container.name"
      to: "@docker-container-name"
    - from: "docker.container.labels.io.rancher.container.name"
      to: "@container-name"
```

Any similar issues with`renaming` configuration duplication?

**Comment by andrewkroh** [comment]
I made a change to fix the duplication of `@timestamp` that I think would help solve parts of this issue. https://github.com/elastic/beats/pull/12449.

The part that it does not solve is that the field being copied into `@timestamp` must have a data type that is a `time.Time`. In most cases you only have a string if manipulating logs. To get around that issue we would need a `date` processor or have the `beat.Event` object try to parse the `@timestamp` values that are strings.

**Comment by HeWhoWas** [comment]
Has there been any progress on this issue since the last comments? I'm running into this - in fact, my log files are logging JSON with the @timestamp key already set - and can't figure out how to fix the timestamp from filebeat without an ingest pipeline.

**Comment by zhisheng17** [comment]
hello，@timestamp can rename to timestamp? @metadata can rename metadata?

**Comment by gszyszka** [comment]
looks like you actually process the `@timestamp` field in filebeat: https://www.elastic.co/guide/en/beats/filebeat/master/processor-timestamp.html

**Comment by Jim-Ivey** [comment]
I'm trying to use the `timestamp` processor in a beat module ingestion without success.  

Here's the relevant part of my pipeline:

```
    {
        "timestamp": {
            "field": "timestamp",
            "layouts": ["2006-01-02T15:04:05.999+0700"]
        }
    }
```

And I get the error, `No processor type exists with name [timestamp]`.

I set up my dev environment in the manner described [here](https://www.elastic.co/guide/en/beats/devguide/master/beats-contributing.html#setting-up-dev-environment) within the last couple weeks.  So, I believe I have the most current version of everything.

Bottom line:  I want `@timestamp` to represent the timestamp parsed from the log and not the time the log was parsed.  Any clues for how to do that would be greatly appreciated.

**Comment by andrewkroh** [comment]
`No processor type exists with name` sounds like an error from Elasticsearch Ingest Node which has a [date](https://www.elastic.co/guide/en/elasticsearch/reference/7.5/date-processor.html) processor. Where are you configuring this? As a processor in your beat's yml config file or in an Elasticsearch Ingest Node pipeline.

The `timestamp` processor is part of [Beats processors](https://www.elastic.co/guide/en/beats/filebeat/current/defining-processors.html).

**Comment by Jim-Ivey** [comment]
In a beat's module's ingestion pipeline: e.g., `filebeat/module/my_mod/my_fileset/ingest/pipeline.json`.

**Comment by andrewkroh** [comment]
That file contains the Ingest Node Pipeline that is installed to Elasticsearch. Beats processors can be used in the config.yml only. And Elasticsearch processors only in the ingest/pipeline.json.

**Comment by fzyzcjy** [comment]
Hi will this feature to be allowed? Because ingest seems to be a bit too heavy, and I want to directly use `rename`/`drop_fields` to do so.

Thanks!

**Comment by fzyzcjy** [comment]
ok I see:

```
        - timestamp:
            # ... By default the timestamp processor writes the parsed result to the @timestamp field. ...
            # https://www.elastic.co/guide/en/beats/filebeat/current/processor-timestamp.html
            field: 'front_log.timestamp'
            layouts:
              - '2006-01-02T15:04:05.999999Z'
            test:
              - '2021-03-25T06:40:58.498982Z'
        - drop_fields:
            fields:
              - 'front_log.timestamp'
```
